File: ropchain/gadgets/asm/xchg.py
from ropchain import ropchain
from ropchain.gadgets import util, gadget

def find(r1, r2, gadgets, canUse):
    rop, canUse = gadget.find(gadgets, canUse, 'xchg', r1, r2)
    if rop == None:
        rop, canUse = gadget.find(gadgets, canUse, 'xchg', r2, r1)
        return util.optROPChain(rop)
    else:
        return ropchain.ROPChain(rop)

File: ropchain/lib/src/asm/xchg.cpp
#include "xchg.h"

OptROP Xchg::find(const Opcode& op1, const Opcode& op2,
        const Gadgets& gadgets, RegSet& aval) {
    if(const auto g1 = Util::find(gadgets, aval, "xchg", op1, op2)) {
        return ROPChain(g1.value());
    }
    if(const auto g2 = Util::find(gadgets, aval, "xchg", op2, op1)) {
        return ROPChain(g2.value());
    }
    return {};
}
